### PR TITLE
add request retries for non-404 errors

### PR DESCRIPTION
**Pull request by jakubgs**
Because some people are on very shitty connections.

### CODE REVIEW

File: fetcher/fetcher.go [hunk 1]
```diff
@@ -61,24 +66,34 @@ func NewPool(limit, timeout int, repos []string) Pool {
...
-	client := &http.Client{
-		Timeout: time.Duration(p.timeout) * time.Second,
-	}
-	resp, err := client.Get(url)
-	if err != nil {
-		return nil, err
-	}
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("failed to fetch with: %d", resp.StatusCode)
+func (p *Pool) RetryFetch(url string) (io.ReadCloser, error) {
```
Comment (Samyoul) [inline fetcher/fetcher.go]: Would there be a case where Fetching does not use retry functionality?

Would this function ever be called outside of the Pool?

File: fetcher/fetcher.go [hunk 2]
```diff
@@ -61,24 +66,34 @@ func NewPool(limit, timeout int, repos []string) Pool {
...
+		if err != nil {
+			return nil, err
+		}
+		if resp.StatusCode == http.StatusOK {
+			return resp.Body, nil
+		}
+		/* 404 must mean it doesn't exist. */
+		if resp.StatusCode == http.StatusNotFound {
+			break
+		}
```
Comment (Samyoul) [inline fetcher/fetcher.go:85]: Do you want to tell other routines that this URL gives you a 404? Should this URL be ignored if found again?

I think this functionality already exists in this code, does the application record the URL before or after the URL has been parsed?

Comment (jakubgs) [reply fetcher/fetcher.go:85]: * the error is returned below
* the caching of urls is done in the finder
* record how and what for?

File: fetcher/fetcher.go (hunk 1 above)
Comment (jakubgs) [reply fetcher/fetcher.go]: I don't see why it would be.

File: fetcher/fetcher.go (hunk 2 above)
Comment (Samyoul) [reply fetcher/fetcher.go:85]: Just to prevent another routine from hitting the 404.

File: fetcher/fetcher.go (hunk 1 above)
Comment (Samyoul) [reply fetcher/fetcher.go]: Should `RetryFetch` be unexported?

**Review by Samyoul — APPROVED** [review]
@jakubgs I've gone over the changes in here and to be honest I don't see anything that is an issue. You've implemented a good solution for retrying dodgy connections.

I've added a few points to the code but those are just little concerns.

File: fetcher/fetcher.go (hunk 1 above)
Comment (jakubgs) [reply fetcher/fetcher.go]: Yeah, that's a good point, it should.

File: fetcher/fetcher.go (hunk 2 above)
Comment (jakubgs) [reply fetcher/fetcher.go:85]: Right, that's already handled by:
https://github.com/status-im/go-maven-resolver/blob/64a31400ba0a4950392b8fb0aca2ab6b8220cfd0/finder/finder.go#L99-L102
So we're good.

File: fetcher/fetcher.go (hunk 1 above)
Comment (jakubgs) [reply fetcher/fetcher.go]: Same for `TryRepos` and `TryRepo` I guess.

**Pull request merged**